sklearn/linear_model/ridge.py

```diff
@@ -1087,55 +1361,49 @@ def fit(self, X, y, sample_weight=None):
             X, y, self.fit_intercept, self.normalize, self.copy_X,
             sample_weight=sample_weight)
 
-        gcv_mode = self.gcv_mode
-        with_sw = len(np.shape(sample_weight))
-
-        if gcv_mode is None or gcv_mode == 'auto':
-            if sparse.issparse(X) or n_features > n_samples or with_sw:
-                gcv_mode = 'eigen'
-            else:
-                gcv_mode = 'svd'
-        elif gcv_mode == "svd" and with_sw:
-            # FIXME non-uniform sample weights not yet supported
-            warnings.warn("non-uniform sample weights unsupported for svd, "
-                          "forcing usage of eigen")
-            gcv_mode = 'eigen'
+        gcv_mode = _check_gcv_mode(X, self.gcv_mode)
 
         if gcv_mode == 'eigen':
-            _pre_compute = self._pre_compute
-            _errors = self._errors
-            _values = self._values
+            _decompose = self._decompose_gram
+            _errors = self._errors_gram
+            _values = self._values_gram
         elif gcv_mode == 'svd':
             # assert n_samples >= n_features
-            _pre_compute = self._pre_compute_svd
-            _errors = self._errors_svd
-            _values = self._values_svd
-        else:
-            raise ValueError('bad gcv_mode "%s"' % gcv_mode)
+            if sparse.issparse(X):
+                _decompose = self._decompose_covariance_sparse
+                _errors = self._errors_covariance_sparse
+                _values = self._values_covariance_sparse
+            else:
+                _decompose = self._decompose_covariance_dense
+                _errors = self._errors_covariance_dense
+                _values = self._values_covariance_dense
 
         if sample_weight is not None:
             X, y = _rescale_data(X, y, sample_weight)
-
-        centered_kernel = not sparse.issparse(X) and self.fit_intercept
-
-        v, Q, QT_y = _pre_compute(X, y, centered_kernel)
+            self._sqrt_sw = np.sqrt(sample_weight)
+            self._weight_sum = sample_weight.sum()
+            self._with_sw = True
+        else:
+            self._with_sw = False
+            self._sqrt_sw = np.ones(X.shape[0], dtype=X.dtype)
+            self._weight_sum = float(X.shape[0])
+        self._normalized_sqrt_sw = self._sqrt_sw / np.linalg.norm(
+            self._sqrt_sw)
+        self._sqrt_sw_matrix = sparse.dia_matrix(
+            (self._sqrt_sw, 0), shape=(X.shape[0], X.shape[0]))
```

Comment:
I had stored them to avoid having long function signatures, at the moment they
are being removed at the end of the fit. we could also put the rest of fit in a
try/except block to make sure they get cleaned up even if `fit` fails.

but passing them as arguments definitely makes sense, I will make the change
right away.

Reply:
Yes I don't like the statefulness and implicitness of passing those as attributes + cleanup.

Reply:
almost done, still need to get rid of `_X_offset`